### PR TITLE
Skip creating Firestore indexes if already added

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:
This PR also changes the indexes JSON file to our own format, which means dependency on the `firebase` command will be removed.
The reason why it did that is it became necessary to parse indexes inside Ops container in order to check indexes already exists. Besides, the indexes emitted by `gcloud firestore indexes composite list` includes project-specific value, that’s why we settled on using our own type.

Also unifying the parsing way makes it easier to solve https://github.com/pipe-cd/pipe/issues/1564. I'm looking into publishing the parsing APIs once got merged.

FYI: I've confirmed it works as well as we'd like this time as well.

**Which issue(s) this PR fixes**:

Fixes https://github.com/pipe-cd/pipe/issues/1559

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/ops/firestoreindexensurer/indexes.json [hunk 1]
```diff
@@ -1,209 +1,306 @@
...
+        "order": "ASCENDING",
+        "arrayConfig": ""
+      },
+      {
+        "fieldPath": "UpdatedAt",
+        "order": "DESCENDING",
+        "arrayConfig": ""
+      },
+      {
+        "fieldPath": "__name__",
```
Comment (nghialv) [inline pkg/app/ops/firestoreindexensurer/indexes.json:17]: Do you know why was this included?

File: pkg/app/ops/firestoreindexensurer/gcloud.go [hunk 2]
```diff
@@ -92,19 +94,59 @@ func (c *gcloud) createIndex(ctx context.Context, idx *index) error {
...
+	indexes := make([]index, 0, len(rawIndexes))
+	for _, idx := range rawIndexes {
+		// Supposed to be like "projects/project-a/databases/(default)/collectionGroups/CollectionA/indexes/CICAgLjRnZMK"
+		name := strings.Split(idx.Name, "/")
+		if len(name) < 6 {
+			c.logger.Warn("index has unexpected name", zap.String("name", idx.Name))
+			continue
+		}
+		indexes = append(indexes, index{
+			CollectionGroup: name[5],
```
Comment (nghialv) [inline pkg/app/ops/firestoreindexensurer/gcloud.go:134]: I am concerned about using only the collection name to handle the index.
What happens if our users have other collections with the same name at other paths?
Instead, Do we have any way to use the full path as the collection ID?

File: pkg/app/ops/firestoreindexensurer/indexes.json (hunk 1 above)
Comment (nakabonne) [reply pkg/app/ops/firestoreindexensurer/indexes.json:17]: Actually, this `__name__` field is automatically added for all indexes when it wasn't explicitly given.

File: pkg/app/ops/firestoreindexensurer/gcloud.go (hunk 2 above)
Comment (nakabonne) [reply pkg/app/ops/firestoreindexensurer/gcloud.go:134]: Thank you. At first, we can specify only collection groups like `Application`, `Deployment`, etc. (https://cloud.google.com/sdk/gcloud/reference/firestore/indexes/composite/create)
That is, we cannot create indexes for each paths. The only one collection group with the name `Application` can exist.

Comment (nakabonne) [reply pkg/app/ops/firestoreindexensurer/gcloud.go:134]: >Do we have any way to use the full path as the collection ID?

Does it mean we change the collection ID in our GCP project for dev to other one? (kind of like PipeCD/Application)

File: pkg/app/ops/firestoreindexensurer/indexes.json (hunk 1 above)
Comment (nghialv) [reply pkg/app/ops/firestoreindexensurer/indexes.json:17]: Yes, I know. Just want to know why they were created.
👍

Comment (nakabonne) [reply pkg/app/ops/firestoreindexensurer/indexes.json:17]: Ah, okay. I don't know why it does exactly. I was looking for it but the doc says just "it's required".

File: pkg/app/ops/firestoreindexensurer/gcloud.go (hunk 2 above)
Comment (nghialv) [reply pkg/app/ops/firestoreindexensurer/gcloud.go:134]: > The only one collection group with the name Application can exist.

Actually, many collections can have the same name as sub-collection and we are using that feature too. You can check the dev environment to see.

> Does it mean we change the collection ID in our GCP project for dev to other one? (kind of like PipeCD/Application)

For example, in a single GCP project, we have multiple pipecd environments (dev, prod) (or db for pipecd and db for user's project) and many collections can have the same name.

The collection path will be something like: `/pipecd/dev/Application`.

From gcloud doc:
> --collection-group=COLLECTION_GROUP
ID of the collection group or fully qualified identifier for the collection group. To set the [collection-group] attribute: provide the argument [--collection-group] on the command line.

Not sure what is the format of the "fully qualified identifier". But the path I wanted to say is this "fully qualified identifier".

Comment (nakabonne) [reply pkg/app/ops/firestoreindexensurer/gcloud.go:134]: >Actually, many collections can have the same name as sub-collection and we are using that feature too.

You're totally right. But what I'm trying to say is about a "collection group", not a collection. I assume the collection group is unique inside a GCP project. And I assume our `Application` collection belongs to the collection group named `Application`.

Maybe I got wrong. It might be easy to talk via Zoom 😄

**Comment by khanhtc1202** [comment]
🏃‍♂️ 
/approve
